source/containerimage/pull.go

```diff
@@ -160,45 +168,89 @@ func mainManifestKey(ctx context.Context, desc specs.Descriptor, platform specs.
 	return digest.FromBytes(dt), nil
 }
 
-func (p *puller) CacheKey(ctx context.Context, g session.Group, index int) (string, bool, error) {
+func (p *puller) CacheKey(ctx context.Context, g session.Group, index int) (_ string, cacheOpts solver.CacheOpts, _ bool, err error) {
 	if p.Puller.Resolver == nil {
 		p.Puller.Resolver = pull.NewResolver(g, p.ResolverOpt)
 	} else {
 		p.ResolverOpt.Auth.AddSession(g)
 	}
-	_, desc, err := p.Puller.Resolve(ctx)
-	if err != nil {
-		return "", false, err
-	}
-	if index == 0 || desc.Digest == "" {
-		k, err := mainManifestKey(ctx, desc, p.Platform)
+
+	p.cacheKeyOnce.Do(func() {
+		ctx, done, err := leaseutil.WithLease(ctx, p.LeaseManager, leases.WithExpiration(5*time.Minute), leaseutil.MakeTemporary)
 		if err != nil {
-			return "", false, err
+			p.cacheKeyErr = err
+			return
+		}
+		p.releaseTmpLeases = done
+		imageutil.AddLease(p.releaseTmpLeases)
+		defer func() {
+			if p.cacheKeyErr != nil {
+				p.releaseTmpLeases(ctx)
+			}
+		}()
+
+		resolveProgressDone := oneOffProgress(ctx, "resolve "+p.Src.String())
+		defer func() {
+			resolveProgressDone(err)
+		}()
+
+		p.manifest, err = p.PullManifests(ctx)
+		if err != nil {
+			p.cacheKeyErr = err
+			return
+		}
+
+		if len(p.manifest.Remote.Descriptors) > 0 {
+			pw, _, _ := progress.FromContext(ctx)
+			progressController := &controller.Controller{
+				Writer: pw,
+			}
+			if p.vtx != nil {
+				progressController.Digest = p.vtx.Digest()
+				progressController.Name = p.vtx.Name()
+			}
+
+			descHandler := &cache.DescHandler{
+				Provider: p.manifest.Remote.Provider,
+				ImageRef: p.manifest.Ref,
```

Comment:
Was looking at this for some resolver changes with `session.Group` and wondered if we really need `ImageRef` at all. Afaics this is used in 2 places, cross-repo annotation and unlazy description message. For both, I think we should just pass them with `GetByBlob` and keep in metadata?

Reply:
My original concern with that was the case where there's multiple images identical in terms of content but using different image refs. In theory, we could store a list of `ImageRef`s for the blob in the metadata, which I think would work fine with the cross-repo stuff since you can add multiple distribution source annotations.

The only part I'm not sure about is the unlazy description message; if we have a list of `ImageRef`s we wouldn't know which one is actually associated with the provider, so if we just chose one it might not actually be the repo we're really pulling it from. Do you think that matters? AFAIK that description message isn't important for functionality, though it would be odd for it to say it pulled the ref from a place it didn't actually do the pull from.

Reply:
> Do you think that matters?

I don't think it does. The same behavior already exists today. If you pull the same image from 2 refs, one of them will win and another will be skipped as already exists. When you are pulling you can't control what side your current pull will land.

Reply:
Okay cool, I made the update to remove `ImageRef` as a separate commit because it was a little more involved than I originally thought, let me know what you think.